lib/core/src/Cardano/Wallet/Api/Server.hs

```diff
@@ -3478,8 +3478,16 @@ instance IsServerError ErrBalanceTx where
                 [ "The transaction is already balanced. "
                 , "Please send a transaction that requires more inputs/outputs to be picked to be balanced."
                 ]
-        ErrBalanceTxUpdateError (ErrUpdateSealedTxBodyError hint) ->
-            apiError err500 CreatedInvalidTransaction hint
+        ErrBalanceTxUpdateError ErrByronTxNotSupported ->
+            apiError err403 CreatedInvalidTransaction
+                "Balancing Byron transactions is not supported."
+        ErrBalanceTxUpdateError (ErrExistingKeyWitnesses n) ->
+            apiError err403 CreatedInvalidTransaction $ mconcat
+                [ "The transaction could not be balanced, because it contains "
+                , T.pack (show n), " "
+                , "existing key-witnesses which would be invalid after "
```

Comment:
I would even say that at this stage we should not have any witnesses collected. signing should be done only after tx is balanced (hence I slightly incline to introduce AlreadyBalancedTx error in balance tx PR)

Reply:
I added a `Please sign the transaction after it is balanced instead.` line afterwards.

However, if for some reason a user want to add key-witness, I'm not sure if the balance-endpoint can take the size of the witness into account for the coin selection (if it is not known to the wallet). Maybe it does…